File: 243_Shortest_Word_Distance.java
/*
https://leetcode.com/problems/shortest-word-distance/
Given a list of words and two words word1 and word2, return the shortest distance between these two words in the list.

Store that last index of each word. Update result with min of index difference and previous result.
Time - O(n)
Space - O(1)
*/

public class Solution {
    public int shortestDistance(String[] words, String word1, String word2) {

        int last1 = -1;
        int last2 = -1;
        int minDist = words.length;     // both words are present so minDist must be < length

        for (int i = 0; i < words.length; ++i) {

            if (words[i].equals(word1)) {
                last1 = i;
                if (last2 != -1)
                    minDist = Math.min(minDist, Math.abs(last1 - last2));
            }

            if (words[i].equals(word2)) {
                last2 = i;
                if (last1 != -1)
                    minDist = Math.min(minDist, Math.abs(last1 - last2));
            }
        }
        return minDist;
    }
}

File: python_1_to_1000/243_Shortest_Word_Distance.py
_author_ = 'jake'
_project_ = 'leetcode'

# https://leetcode.com/problems/shortest-word-distance/
# Given a list of words and two words word1 and word2, return the shortest distance between these two words in the list.
# For example,
# Assume that words = ["practice", "makes", "perfect", "coding", "makes"].
# Given word1 = “coding”, word2 = “practice”, return 3.
# Given word1 = "makes", word2 = "coding", return 1.
# You may assume that word1 does not equal to word2, and word1 and word2 are both in the list.

# Iterate over list, updating the last seen indices whenever word matches word1 or word2. Also update the shortest
# distance.
# Time - O(n)
# Space - O(1)

class Solution(object):
    def shortestDistance(self, words, word1, word2):
        """
        :type words: List[str]
        :type word1: str
        :type word2: str
        :rtype: int
        """
        shortest = len(words)
        i_1, i_2 = float("-inf"), float("-inf")         # last seen indices of word1 and word2

        for i, word in enumerate(words):

            if word == word1:
                i_1 = i
                shortest = min(shortest, i_1 - i_2)
            if word == word2:
                i_2 = i
                shortest = min(shortest, i_2 - i_1)

        return shortest